src/Uuid/Uuid.php

```diff
@@ -0,0 +1,318 @@
+<?php
+
+namespace Kirby\Uuid;
+
+use Closure;
+use Generator;
+use Hidehalo\Nanoid\Client as Nanoid;
+use InvalidArgumentException;
+use Kirby\Cms\Collection;
+use Kirby\Cms\File;
+use Kirby\Cms\Page;
+use Kirby\Cms\Site;
+use Kirby\Cms\User;
+use Kirby\Toolkit\Str;
+
+/**
+ * The `Uuid` classes provide an interface to connect
+ * indetifiable models (page, file, site, user, blocks,
+ * structure entries) with a dedicated UUID string.
+ * It also provides methods to cache these connections
+ * for faster lookup.
+ *
+ * ```
+ * // get UUID string
+ * $model->uuid()->render();
+ *
+ * // get model from an UUID string
+ * Uuid::for('page://HhX1YtRR2ImG6h4')->resolve();
+ *
+ * // cache actions
+ * $model->uuid()->populate();
+ * $model->uuid()->clear();
+ * ```
+ *
+ * @package   Kirby Uuid
+ * @author    Nico Hoffmann <nico@getkirby.com>
+ * @link      https://getkirby.com
+ * @copyright Bastian Allgeier
+ * @license   https://getkirby.com/license
+ */
+class Uuid
+{
+	protected const TYPE = 'uuid';
+
+	/**
+	 * Customisable callback function for generating new ID strings
+	 * instead of using Nanoid. Receives length of string as parameter.
+	 */
+	public static Closure|null $generator = null;
+
+	public Collection|null $context;
+	public Identifiable|null $model;
+	public Uri|null $uri;
+
+	public function __construct(
+		string|null $uuid = null,
+		Identifiable|null $model = null,
+		Collection|null $context = null
+	) {
+		$this->context = $context;
+		$this->model   = $model;
+
+		if ($model) {
+			$this->uri = new Uri([
+				'scheme' => static::TYPE,
+				'host'   => static::retrieveId($model)
+			]);
+		} elseif ($uuid) {
+			$this->uri = new Uri($uuid);
+		}
```

Comment:
In theory both `$uuid` and `$model->uuid()` should be the same and also in theory one should never construct a `Uuid` object with both a UUID string and a model object, but what happens if both are passed and there is a mismatch? Maybe we should throw an exception if both are passed and don't match.